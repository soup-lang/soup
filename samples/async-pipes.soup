list: [4, 2, 3, 1]

func print(element, cb)
    setTimeout(%, 0)
    console.log('Element: ' + element)
    cb()

func echo(cb, index)
    setTimeout(%, 0)
    console.log('Index: ' + index)
    cb()

func g(p, q, cb)
    setTimeout(cb, 0)
    return 0

x: list |: g(print($, %), echo(%, $i), %)
console.log(x)

func asyncSqr(element, cb)
    setTimeout(%, 0)
    cb(element * element)

console.log(list |:
    asyncSqr($, %x)
    $result.push(x)
)

console.log(list |:
    asyncSqr($, %x)
    if x % 3 = 1
        $r.push(x)
)

list |:
    console.log($i + ':' + $)
    setTimeout(%, 0)

console.log('-- references in nested --')

func nestedPiping(x, %%)
    x |:
        i: $
        x |:
            setTimeout(%, 0)
            console.log('+', i + $)
        console.log('.', $, $i)

nestedPiping([0, 1], %%)
nestedPiping('ab', %%)

console.log('---- async on object ----')

obj: {fox: 'brown quick', dog: 'lazy'}

func asyncIdent(x, %%)
    setTimeout(%, 0)
    return x

console.log((obj |: asyncIdent($, %%)).sort())
console.log(obj |: asyncIdent($i, %%))
console.log((obj |: asyncIdent($k, %%)).sort())

console.log('End')
