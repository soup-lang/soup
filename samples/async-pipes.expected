Element: 4
Index: 0
Element: 2
Index: 1
Element: 3
Index: 2
Element: 1
Index: 3
[ 0, 0, 0, 0 ]
[ 16, 4, 9, 1 ]
[ 16, 4, 1 ]
0:4
1:2
2:3
3:1
-- references in nested --
+ 0
+ 1
. 0 0
+ 1
+ 2
. 1 1
+ aa
+ ab
. a 0
+ ba
+ bb
. b 1
---- async on object ----
[ 'brown quick', 'lazy' ]
[ 0, 1 ]
[ 'dog', 'fox' ]
End
